controllers/topology/internal/mergepatch/managed_paths_test.go

```diff
@@ -0,0 +1,296 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package mergepatch
+
+import (
+	"testing"
+
+	. "github.com/onsi/gomega"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	"sigs.k8s.io/cluster-api/controllers/topology/internal/contract"
+)
+
+func Test_getManagedPaths(t *testing.T) {
+	tests := []struct {
+		name string
+		obj  client.Object
+		want []contract.Path
+	}{
+		{
+			name: "Return no paths if the annotation does not exists",
+			obj: &unstructured.Unstructured{
+				Object: map[string]interface{}{},
+			},
+			want: []contract.Path{},
+		},
+		{
+			name: "Return paths from the annotation",
+			obj: &unstructured.Unstructured{
+				Object: map[string]interface{}{
+					"metadata": map[string]interface{}{
+						"annotations": map[string]interface{}{
+							clusterv1.ClusterTopologyManagedFieldsAnnotation: "foo, bar.baz",
+						},
+					},
+				},
+			},
+			want: []contract.Path{
+				{"spec", "foo"},
+				{"spec", "bar", "baz"},
+			},
+		},
+		{
+			name: "Handle label names properly",
+			obj: &unstructured.Unstructured{
+				Object: map[string]interface{}{
+					"metadata": map[string]interface{}{
+						"annotations": map[string]interface{}{
+							clusterv1.ClusterTopologyManagedFieldsAnnotation: "bar.foo%bar%baz, foo",
+						},
+					},
+				},
+			},
+			want: []contract.Path{
+				{"spec", "bar", "foo.bar.baz"},
+				{"spec", "foo"},
+			},
+		},
+		{
+			name: "Handle deep nesting ",
+			obj: &unstructured.Unstructured{
+				Object: map[string]interface{}{
+					"metadata": map[string]interface{}{
+						"annotations": map[string]interface{}{
+							clusterv1.ClusterTopologyManagedFieldsAnnotation: "kubeadmConfigSpec.clusterConfiguration.imageRepository, " +
```

Comment:
Should we consider using some sort of encoding for this annotation? The paths can be easily compressed, and we could b64 encode the output

Reply:
I kind of like the idea of this being human readable, so was considering something like 
`{ kubeadmConfigSpec { clusterConfiguration { imageRepository, version }, initConfiguration { ... } } }`
But this requires a little encoder/parser and I did not include in this first iteration, but I really want to sort it out before the actual release